API Implement many_many through

tractorcow opened the pull request:
Fixes #5615

Merge with CMS changes: https://github.com/silverstripe/silverstripe-cms/pull/1639

This change introduces a slight refactor in model relationship handling; Validation has been updated, and moved to DataObjectSchema.

The basic implementation of this feature is that a many_many relation now utilises a nested has_many relation, which is rewritten as a joined subquery within it in place of a join table.

In order to ensure that both has_many and many_many queries are joined, and independently augmented correctly, a ManyManyThroughQueryManipulator object has been implemented to merge them during `getFinalisedQuery()`.

I've also taken advantage of the `::class` language feature in PHP 5.5 to replace some class literals.

Includes the following ORM changes:
- API Remove DataObject::validateModelDefinitions, and move to DataObjectSchema
- API Remove deprecated 3.0 syntax for addSelect()
- API made DataList::createDataObject public
- API Move component parsing logic to DataObjectSchema
- API Remove support for triangular has_many / belongs_many relationships

kinglozzer commented on ORM/DataObject.php:
```diff
@@ -2149,78 +2069,27 @@ public function manyMany() {
-	 * Return information about a specific many_many component. Returns a numeric array of:
+	 * Return information about a specific many_many component. Returns a numeric array.
+	 * The first item in the array will be the class name of the relation: either
+	 * RELATION_MANY_MANY or RELATION_MANY_MANY_THROUGH constant value.
```
Can’t find these constants, doesn’t it just return the FQCN?

tractorcow replied:
Sorry, those were added but then removed. I'll update the doc.

tractorcow replied:
Done!

kinglozzer commented on ORM/DataObjectSchema.php:
```diff
@@ -367,4 +368,396 @@ public function classForField($candidateClass, $fieldName)  {
+			return null;
+		}
+		foreach ($otherManyMany as $inverseComponentName => $nextClass) {
+			if ($nextClass === $parentClass) {
```
The logic has changed here slightly - it used to also check `|| is_subclass_of($parentClass, $nextClass)`. I’ve run through it mentally a few times and I’m sure the removal of that logic is fine, just noting it :)

tractorcow replied:
Yeah, supporting for what I called "triangular relationships" was flakey, and led to some weird behaviour infering reciprocal relationships.

kinglozzer commented on ORM/ManyManyList.php:
```diff
@@ -210,23 +211,30 @@ public function add($item, $extraFields = array()) {
 		$foreignIDs = $this->getForeignID();
 		if(empty($foreignIDs)) {
-			throw new Exception("ManyManyList::add() can't be called until a foreign ID is set", E_USER_WARNING);
+			throw new BadMethodCallException("ManyManyList::add() can't be called until a foreign ID is set", E_USER_WARNING);
```
Can we remove the `E_USER_WARNING`?

tractorcow replied:
Ah yes that should not be there.

tractorcow replied:
Done!

kinglozzer commented on ORM/ManyManyThroughList.php:
```diff
@@ -0,0 +1,189 @@
+		// Validate foreignID
+		$foreignIDs = $this->getForeignID();
+		if (empty($foreignIDs)) {
+			throw new BadMethodCallException("ManyManyList::add() can't be called until a foreign ID is set", E_USER_WARNING);
```
`ManyManyThroughList`, and `E_USER_WARNING` to be removed?

tractorcow replied:
Will do.

tractorcow replied:
Done!

kinglozzer commented on docs/en/02_Developer_Guides/00_Model/02_Relations.md:
```diff
@@ -213,50 +213,139 @@ This is not mandatory unless the relationship would be otherwise ambiguous.
 
 	:::php
 	$team = Team::get()->byId(1);
+	$supporters = $team->Supporters()->where(['"Join"."Ranking"' => 1]);
```
Are we able to use the join table name directly here? To use the physician/patient example from [Ruby on Rails](http://guides.rubyonrails.org/association_basics.html#the-has-many-through-association):

```php
$patientsToday = $physician->Patients()->where(['"Appointment"."Date"' => date('Y-m-d')]);
```

Is that possible, or do we have to use the `Join` alias?

tractorcow replied:
Hm, possibly, although it would make the api more complex.

tractorcow replied:
On second thoughts, I think it's worth exploring. I'll give it a try. :)

tractorcow replied:
Done!

kinglozzer commented on docs/en/02_Developer_Guides/00_Model/02_Relations.md:
```diff
@@ -213,50 +213,139 @@ This is not mandatory unless the relationship would be otherwise ambiguous.
-	$supporters = $team->Supporters();
-	// returns a 'ManyManyList' instance.
 
+Note: ->filter() currently does not support joined fields natively.
```
:( will this be possible in future?

tractorcow replied:
Unfortunately in order to support arbitrary augmentSQL'd join queries, the join is wrapped within a subquery with an alias, so it would be quite difficult for the datalist api to infer from the nested relationship robustly. It would essentially require this solution to be re-implemented with some unknown method.

kinglozzer commented:
Nice work! I think we’ll need consensus on the syntax for declaring the many-many-through relationships, but I like the approaches you’ve used for tackling this

tractorcow commented:
@kinglozzer I've responded to your feedback.

- Join alias is now the join table name, not just "Join". '$Join' can still be used in templates as well as the class name, however.
- I've added some extra validation to handle clashes between the new alias and other fields on either side of the relation.
- I've moved DataObject::db() into DataObjectSchema()
- I've completely removed the constant that used to be the join alias. It's now a method `ManyManyThroughQueryManipulator::getJoinAlias()`
- I've removed all the extra arguments to exception constructors.

kinglozzer commented:
Great, thanks! I’ve checked the changes out locally and tested them, and everything works how I was expecting. The lack of `filter()` support is a bummer, but your explanation makes sense - I did play around a little trying to find a solution, but encountered the exact problems you described.

I’m happy with these changes, so I’m going to hit the magic approve button. I won’t merge yet, as I’d like some more eyes on this first :)

tractorcow commented:
@kinglozzer : @sminnee has looked over this and said that it would be fine if you were to take over merging. ;)

sminnee commented:
I haven't had a detailed read yet but I'll do so today click a magic approve button when I'm ready.

tractorcow commented:
Thanks guys. :)

sminnee commented on tests/model/ManyManyThroughListTest.php:
```diff
@@ -0,0 +1,275 @@
+		ManyManyThroughListTest_VersionedObject::class,
+	];
+
+	public function testSelectJoin() {
```
Add a test for sorting by `SortOrder`. Sorting by a field on the join object is a pretty critical feature.

tractorcow replied:
Done.

tractorcow commented on ORM/DataObject.php:
```diff
@@ -1682,8 +1700,7 @@ public function getRelationType($component) {
 	public function inferReciprocalComponent($remoteClass, $remoteRelation) {
-		/** @var DataObject $remote */
-		$remote = $remoteClass::singleton();
+		$remote = DataObject::singleton($remoteClass);
```
DataObject::singleton() simple invokes Injector::inst()->get(), but it has `@return static` to enable type-hinting, whereas Injector::inst()->get() has none. I've found that typehinting in code helps uncover bugs a lot easier, so would you agree it's beneficial to leave it?

sminnee replied:
Fair call; fine to leave for now. The lack of type hinting from our DI system can be a problem for another day ;-)

tractorcow commented:
I have a feeling the current getSchema() api might need a little bit of think. Perhaps getSchema() should return a schema object scoped to the given parent, similar to Config_forClass?

tractorcow commented:
I'm going to do a bunch more cleanup of class constants, as well as consolidating some of the DataObjectSchema API and remove methods, as suggested.

I think it's ok for DataObjectSchema to be big, as long as it's well defined, and reduces complexity of DataObject itself; All it does it provide information on how classes, fields, tables, and relations should fit together; DataObject itself still handles physical relationships between instances.

The filtering issue is a tough one and difficult to crack easily in a way that works the way we'd expect it to. The issue I had were to do with the immutability of DataList and dealing with multiple datalists. The analogy I would give is to take a russian doll design pattern and apply it to a binary tree;

The solution to make ->filter() work natively with nested datalists is possible but would just be more complicated. I anticipate that rather than creating the inner HasMany list dynamically, we would need to hold an instance for the lifetime of the outer datalist; Probably with a more complicated alterDataQuery / __clone that would ensure consistency. This is probably something I'll approach when I've more time to come back to it. :D

The downside of not doing this is that you need to write `->where(['"JoinClass"."Field"' => $value])` instead of `->filter('JoinClass.Field', $value)`. It also means needing to express search filters such as BeginsWith / EndsWith etc explicitly via SQL.

tractorcow commented:
@sminnee I've added the sorting unit tests. Additionally in a much larger commit I've moved a large quantity of db schema methods to DataObjectSchema.

I spoke with @chillu earlier and we came up with a more useful version of the `db` method that would allow us to supercede a large number of methods on DataObject, meaning we can simplify the class.

tractorcow commented:
New CMS pr at https://github.com/silverstripe/silverstripe-cms/pull/1639

tractorcow commented on ORM/DataObject.php:
```diff
@@ -3407,10 +3276,8 @@ public function requireTable() {
-		$this->validateModelDefinitions();
 		if($fields) {
-			$hasAutoIncPK = get_parent_class($this) === 'SilverStripe\ORM\DataObject';
+			$hasAutoIncPK = get_parent_class($this) === __CLASS__;
```
Done!

tractorcow commented on ORM/DataObject.php:
```diff
@@ -3421,29 +3288,34 @@ public function requireTable() {
-					$childField = "{$childTable}ID";
+			foreach($manyMany as $component => $spec) {
+				// Get many_many spec
+				$manyManyComponent = $this->getSchema()->manyManyComponent(get_class($this), $component);
```
Done.

tractorcow commented on tests/model/ManyManyThroughListTest.php:
```diff
@@ -0,0 +1,275 @@
+ * @method ManyManyThroughListTest_Object Parent()
+ * @method ManyManyThroughListTest_Item Child()
+ */
+class ManyManyThroughListTest_JoinObject extends DataObject implements TestOnly
```
Done.

tractorcow commented on ORM/DataObject.php:
```diff
@@ -270,88 +283,7 @@ class DataObject extends ViewableData implements DataObjectInterface, i18nEntity
-	 * @param boolean $aggregated True if parent classes should be checked, or false to limit to this class
-	 * @return string|false Class spec name of composite field if it exists, or false if not
-	 */
-	public static function is_composite_field($class, $name, $aggregated = true) {
```
The removal of all of these static methods to a non-static DataObjectSchema service means we can now DI our schema. :)

tractorcow commented on ORM/DataObjectSchema.php:
```diff
@@ -137,6 +138,62 @@ public function baseDataTable($class) {
+	 *  - `includeClass`: If true prefix the field specification with the class name in RecordClass.Column(spec) format.
+	 * @return array
+	 */
+	public function fieldSpecs($classOrInstance, $options = []) {
```
This is what was DataObject::db()

kinglozzer replied:
Being really nitpicky but... could we use a bitmask instead of an array of case-sensitive strings for passing these options around? Personal preference, so feel free to say no if you disagree!

sminnee replied:
Agree with @kinglozzer: a bitmask based on class constants would be less error-prone and more easily self-documenting.

tractorcow replied:
https://github.com/silverstripe/silverstripe-framework/pull/6161

tractorcow commented on ORM/DataObjectSchema.php:
```diff
@@ -212,15 +269,48 @@ protected function buildTableName($class) {
 	 * @return array Map of fieldname to specification, similiar to {@link DataObject::$db}.
 	 */
-	public function databaseFields($class) {
+	public function databaseFields($class, $aggregated = true) {
```
Added $aggregate to make this more symmetrical with compositeFields()

tractorcow commented:
Oh finally, green CI. :)

kinglozzer commented on ORM/DataObjectSchema.php:
```diff
@@ -347,24 +450,476 @@ public function tableForField($candidateClass, $fieldName) {
+	/**
+	 * Return information about a specific many_many component. Returns a numeric array.
+	 * The first item in the array will be the class name of the relation: either
+	 * RELATION_MANY_MANY or RELATION_MANY_MANY_THROUGH constant value.
```
Looks like this docblock was missed earlier - those constants were removed

tractorcow replied:
Must have been a rebase issue...

kinglozzer commented on ORM/DataObject.php:
```diff
@@ -410,13 +342,13 @@ public function __construct($record = null, $isSingleton = false, $model = null,
+			// Get all field specs scoped to class for later lazy loading
+			$fields = static::getSchema()->fieldSpecs(static::class, ['includeClass', 'dbOnly']);
+			foreach($fields as $field => $fieldSpec) {
+				$fieldClass = strtok($fieldSpec, ".");
```
Why is this necessary? I can’t spot where it came from 😅

kinglozzer commented:
Reviewed the new commit, nothing merge blocking 👍

tractorcow commented:
See https://github.com/silverstripe/silverstripe-framework/pull/6161
